### ISSUE TITLE
early-return false positive

### ISSUE THREAD

**Issue by mdelah** [mgechev/revive#775]
**Describe the bug**
The `early-return` linter can give bad advice if there are multiple non-returning blocks in an if-else chain.

**To Reproduce**
1. I updated revive `go install github.com/mgechev/revive@latest`
2. I run it with the following flags & configuration file:

```shell
revive -config revive.toml
```

```toml
# revive.toml
[rule.early-return]
Disabled = false
```

On the following program:

```go
package main

import "flag"

var (
        A = flag.Bool("A", false, "A flag")
        B = flag.Bool("B", false, "B flag")
)

func main() {
        flag.Parse()
        if *A {
                println("A branch hit")
        } else if *B {
                println("B branch hit")
        } else {
                println("neither A nor B branch hit")
                return
        }
        println("either A or B branch hit")
}
```

The following is output:

```
main.go:14:9: if c {...} else {... return } can be simplified to if !c { ... return } ...
```

**Expected behavior**

No warning is produced.

If we follow the linter's suggestion, we end up with:

```go
if *A {
    println("A branch hit")
} else if !*B {
    println("neither A nor B branch hit")
    return
}
println("B branch hit")
```

Which is a change in behaviour (the program now prints `B branch hit` when the `-A` switch is passed but not `-B`).

To fix the behaviour, we would need to re-introduce a trailing `else`:

```go
if *A {
    println("A branch hit")
} else if !*B {
    println("neither A nor B branch hit")
    return
} else {
    println("B branch hit")
}
```

Which is the same line count as the original code (and somewhat harder to read).

**Desktop**
 - OS: Ubuntu 22.04
 - Go 1.19.3

**Comment by chavacava** [comment]
Hi @mdelah thanks for filling the issue.
My understanding of revive's suggestion is the code could be refactored into:
```go
	if *A {
			println("A branch hit")
	} else if !*B {
			println("neither A nor B branch hit")
			return
		} 
		println("B branch hit")
	}
	
	println("either A or B branch hit")
```

The message 
```
main.go:14:9: if c {...} else {... return } can be simplified to if !c { ... return } ...
```
Makes reference to the `else` of line 14, thus the`if` of line 12.
Do you agree?

**Comment by mdelah** [comment]
Hi @chavacava, thanks for the reply.

Sorry, I'm not sure I see the reference to line 12? The predicate `c` that we are being instructed to invert appears in this case to be `*B`, which sits on line 14, the line indicated in the prompt. Also, the code you have posted is not valid syntax (there is a trailing right-brace).

(Just to be clear, I do like the _idea_ of this rule, and I'd like to enable it on my code base, it's just on constructions of this particular form I think it's not applicable)

**Comment by chavacava** [comment]
@mdelah, you are right, the line 14 is that of the `if *B` (my bad, I've did not count lines correctly :) )
Then, the message of `revive` is pointing to the following code section:
```go
if *B {
  println("B branch hit")
} else {
  println("neither A nor B branch hit")
  return
}
```
if you transform the code as proposed by the message, you will get:
```go
if !*B {
  println("neither A nor B branch hit")
  return
}
println("B branch hit")
```
Then the refactoring of the whole function results in:

```go
func main() {
	flag.Parse()
	if *A {
		println("A branch hit")
	} else {
		if !*B {
			println("neither A nor B branch hit")
			return
		}
		println("B branch hit")
	}
	println("either A or B branch hit")
}
```
```

**Comment by mdelah** [comment]
Yes, that's another way to repair it, but performing such a refactor merely for the sake of placating the linter seems dubious. Consider:

1. The [stated aim](https://github.com/mgechev/revive/blob/master/RULES_DESCRIPTIONS.md#early-return) of the rule is to "minimize nesting statements", but in the end we have strictly increased the nesting (the lines `if !*B {`, `println("neither A nor B branch hit")`, and the following `return` are all now nested 1 level deeper than before, while the other lines remain at their original depth).
2. The overall line count has increased (by 1).
3. I'm not sure about you but as a human I find the new code less natural to read.

Moreover, the refactoring done is _not_ the simple transformation `if c {X} else {Y return}` => `if !c {Y return} X` that was suggested. Some implicit extra steps are needed (either adding additional surrounding braces or re-introducing an else block). An inattentive developer applying the suggested transformation exactly as stated would subtly introduce a bug into their code (I very nearly did this myself!). That's a bit dangerous, no? For me, that alone is reason enough not enable this rule in its current form.

Appreciate that most linting rules (especially more subjective ones) can never be 100% free of false positives, but it seems to me in _any_ if-else chain that has more than one non-returning branch, this rule is not going to be useful (since the code that the rule wants to de-indent would start to get hit by the other non-returning branches if we did). That seems like it should be a relatively easy condition to check, and thus a fairly clear improvement to this rule. What do you think?

**Comment by chavacava** [comment]
The linter makes suggestions is up to the developer to decide what to do. As any static analysis tool, the linter has its limitations. 
In this particular case it could be possible, and maybe not so complex, to avoid the false positive case you spot. Now the problem is to find the pattern the rule must look for to avoid warning in such cases. The pattern must as precise as possible and, at the same time, of an acceptable complexity.

PRs are welcome!

**Comment by mdelah** [comment]
Okay, cheers - I'll see what I can come up with.

**Comment by chavacava** [comment]
PR merged, thanks @mdelah
